### PR TITLE
Implement eth_estimateGas endpoints

### PR DESCRIPTION
**Pull request by atchapcyp**
ref: https://github.com/loomnetwork/loomchain/issues/1483
- [ ] I added unit tests for any code that added
- [ ] I updated the CHANGELOG.md 
- [ ] All IP is original and not copied from another source
- [ ] I assign all copyright to Loom Network for the code in the pull request

### CODE REVIEW

File: e2e/tests/truffle/contracts/SimpleStore.sol [hunk 1]
```diff
@@ -0,0 +1,16 @@
+pragma solidity ^0.5.0;
+
+contract SimpleStore {
```
Comment (enlight) [inline e2e/tests/truffle/contracts/SimpleStore.sol]: Call this `GasEstimateTestContract`

File: e2e/tests/truffle/test/EthFunctions.js [hunk 2]
```diff
@@ -144,5 +145,38 @@ contract('MyToken', async (accounts) => {
     assert.equal(ethOwner.toLowerCase(), web3js.utils.padLeft(owner, 64).toLowerCase(), "result using tokenContract and eth.call");
   });
 
-});
+  it('MyToken eth_estimateGas', async () => {
```
Comment (enlight) [inline e2e/tests/truffle/test/EthFunctions.js]: One test case is enough I think, you can remove this one and keep the one below.

File: evm/evm.go [hunk 3]
```diff
@@ -264,6 +264,35 @@ func (e Evm) GetStorageAt(addr loom.Address, key []byte) ([]byte, error) {
...
+	} else {
+		val = value.Int
+		if val == nil {
+			//there seems like there are serialization issues where we can get bad data here
+			val = common.Big0
+		}
+		if e.validateTxValue && val.Cmp(common.Big0) < 0 {
+			return 0, errors.Errorf("value %v must be non negative", value)
+		}
+	}
```
Comment (enlight) [inline evm/evm.go]: Should be
```golang
if value == nil || value.Int == nil {
  val == common.Big0
} else {
  val = value.Int
  if val.Cmp(common.Big0) < 0 {
    return 0, errors.Errorf("value %v must be non negative", value)
  }
}
```

File: evm/loomevm.go [hunk 4]
```diff
@@ -209,6 +210,8 @@ func (lvm LoomVm) Create(caller loom.Address, code []byte, value *loom.BigUInt)
 }
 
 func (lvm LoomVm) Call(caller, addr loom.Address, input []byte, value *loom.BigUInt) ([]byte, error) {
+	fmt.Println("PASS LOOM VM CALL")
+	defer fmt.Println("END LOOM VM CALL")
```
Comment (enlight) [inline evm/loomevm.go]: Remove these

File: plugin/vm.go [hunk 5]
```diff
@@ -188,6 +189,7 @@ func (vm *PluginVM) StaticCall(caller, addr loom.Address, input []byte) ([]byte,
 }
 
 func (vm *PluginVM) CallEVM(caller, addr loom.Address, input []byte, value *loom.BigUInt) ([]byte, error) {
+	fmt.Println("PASS PLUGIN EVM CALL")
```
Comment (enlight) [inline plugin/vm.go]: Remove

File: rpc/eth/jsonrpc_conversion.go [hunk 6]
```diff
@@ -103,7 +103,7 @@ type JsonTxCallObject struct {
 	GasPrice Quantity `json:"gasPrice,omitempty"`
 	Value    Quantity `json:"value,omitempty"`
 	Data     Data     `json:"data,omitempty"`
-	Nonce    Quantity `json:"nonce.omitempty"`
+	Nonce    Quantity `json:"nonce,omitempty"`
```
Comment (enlight) [inline rpc/eth/jsonrpc_conversion.go]: Huh, can you submit a separate PR for this fix? I want it to go in ASAP

File: rpc/query_server.go [hunk 7]
```diff
@@ -1094,8 +1094,38 @@ func (s *QueryServer) EthGetStorageAt(local eth.Data, position string, block eth
 	return eth.EncBytes(storage), nil
 }
 
-func (s *QueryServer) EthEstimateGas(query eth.JsonTxCallObject) (eth.Quantity, error) {
-	return eth.Quantity("0x0"), nil
+// https://github.com/ethereum/wiki/wiki/JSON-RPC#eth_estimateGas
+func (s *QueryServer) EthEstimateGas(query eth.JsonTxCallObject, block eth.BlockHeight) (resp eth.Quantity, err error) {
+	var caller loom.Address
+	if len(query.From) > 0 {
+		caller, err = eth.DecDataToAddress(s.ChainID, query.From)
```
Comment (enlight) [inline rpc/query_server.go]: Nope, the `from` chain ID should always be assumed to be `eth`.
Yes, `EthCall` does this wrong too, it'll be fixed in another PR I'm working on.

File: rpc/query_server.go [hunk 8]
```diff
@@ -1094,8 +1094,38 @@ func (s *QueryServer) EthGetStorageAt(local eth.Data, position string, block eth
...
+		return resp, err
+	}
+
+	snapshot := s.StateProvider.ReadOnlyState()
+	defer snapshot.Release()
+
+	vm := levm.NewLoomVm(snapshot, nil, nil, nil, true)
+	gasUsed, err := vm.EstimateGas(caller, contract, data, nil)
+	if err != nil {
+		fmt.Printf("error in call estimateGas %+v", err)
```
Comment (enlight) [inline rpc/query_server.go]: This error should be returned

File: rpc/query_server.go [hunk 9]
```diff
@@ -1094,8 +1094,38 @@ func (s *QueryServer) EthGetStorageAt(local eth.Data, position string, block eth
...
+	}
+	fmt.Println("query_server est. gas : ", gasUsed)
```
Comment (enlight) [inline rpc/query_server.go]: Remove this line

**Review by enlight — CHANGES_REQUESTED** [review]

**Comment by eduardonunesp** [comment]
I have tested this branch together with `eth_sendRawTransaction` and the estimated gas makes sense, I'll put a test for it LoomProvider2

File: rpc/query_server.go [hunk 10]
```diff
@@ -1094,8 +1094,38 @@ func (s *QueryServer) EthGetStorageAt(local eth.Data, position string, block eth
...
+
+	data, err := eth.DecDataToBytes(query.Data)
+	if err != nil {
+		return resp, err
+	}
+
+	snapshot := s.StateProvider.ReadOnlyState()
+	defer snapshot.Release()
+
+	vm := levm.NewLoomVm(snapshot, nil, nil, nil, true)
```
Comment (enlight) [inline rpc/query_server.go]: Do we really need debug mode enabled in the EVM to estimate gas?

Comment (enlight) [reply rpc/query_server.go]: Also, if we don't pass the account balance manager to `NewLoomVm` then any call to a payable function, or a function that does a balance check is likely to produce an inaccurate gas estimate since the code path that will be executed is likely to be quite different from the one that will be executed when ETH is available.

File: e2e/tests/truffle/contracts/MyCoin.sol [hunk 11]
```diff
@@ -11,8 +11,37 @@ contract MyCoin is ERC20 {
...
+    function payToSet(uint num) external payable{
+     require(msg.value >= setFee);
+     globalUint = num;
+        emit NewValueSet(num);
+  }
+
+  function getUint() public view returns (uint ){
+      return globalUint;
+  }
+
```
Comment (enlight) [inline e2e/tests/truffle/contracts/MyCoin.sol]: Formatting is messed up in this file, it's time to add a Solidity code formatter.
Yarn install this https://github.com/duaraghav8/Ethlint and set it up in the `e2e/tests/truffle` dir.
Can copy the linter config from the `mainnet` dir in the tg-v2 repo.

**Review by enlight — CHANGES_REQUESTED** [review]

**Comment by atchapcyp** [comment]
rebuild sesame

File: evm/evm.go [hunk 12]
```diff
@@ -264,6 +264,31 @@ func (e Evm) GetStorageAt(addr loom.Address, key []byte) ([]byte, error) {
...
+	contract := common.BytesToAddress(addr.Local)
+	vmenv := e.NewEnv(origin)
+
+	var val *big.Int
+	if value == nil || value.Int == nil {
+		val = common.Big0
+	} else {
+		val = value.Int
+		if val.Cmp(common.Big0) < 0 {
+			return 0, errors.Errorf("value %v must be non negetive", value)
```
Comment (enlight) [inline evm/evm.go]: `non negetive` -> `positive`

**Review by enlight — CHANGES_REQUESTED** [review]

**Comment by atchapcyp** [comment]
rebuild sesame

**Review by enlight — APPROVED** [review]
LGTM
